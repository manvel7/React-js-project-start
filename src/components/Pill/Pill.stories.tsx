import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import Pill, {IProps} from "./Pill"

const meta: Meta<IProps> = {
  title: "Components/Pill",
  component: Pill,
  argTypes: {
    onClick: {action: "onClick"}
  }
}

const PillStory: Story<IProps> = (args) => <Pill {...args}>Pill</Pill>

const Active = PillStory.bind({})
Active.args = {
  isActive: true
}

const NotActive = PillStory.bind({})
NotActive.args = {
  isActive: false
}

const Medium = PillStory.bind({})
Medium.args = {
  size: "md"
}

const Small = PillStory.bind({})
Small.args = {
  size: "sm"
}

export {Active, NotActive, Medium, Small}

export default meta
