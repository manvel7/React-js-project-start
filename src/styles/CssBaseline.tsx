import {
  createGlobalStyle,
  css,
  DefaultTheme,
  GlobalStyleComponent,
  StyledProps
} from "styled-components"
import {mediaBreakpointDown} from "../utils/breakpoint"
import {getValue, toPercent} from "../utils/unit"
import {
  bodyMixin,
  smallMixin,
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin
} from "./typography"

interface IProps {}

const genericResetMixin = css<StyledProps<IProps>>`
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    vertical-align: baseline;
  }

  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }

  ol,
  ul {
    list-style: none;
  }

  blockquote,
  q {
    quotes: none;
  }

  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
`

const appResetMixin = css<StyledProps<IProps>>`
  html {
    /* stylelint-disable property-no-vendor-prefix */
    -ms-overflow-style: scrollbar; /* Set box-sizing so that the width is not affected by padding or border */
    -ms-text-size-adjust: 100%; /* Set default font family in all browsers. */
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); /* Makes 1rem = 10px */
    -webkit-text-size-adjust: 100%; /* Prevent adjustments of font size after orientation changes in IE on Windows Phone and in iOS. */
    /* stylelint-enable property-no-vendor-prefix */
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  body {
    background-color: ${({theme}) => theme.palette.background.white};

    /* Hide overflow content */
    &.overflow {
      overflow: hidden;
    }

    /* Prevent focus outline on elements that can't be accessed via keyboard. */
    [tabindex="-1"]:focus {
      outline: none !important;
    }

    /* Hide overflow in IE */
    svg:not(:root) {
      overflow: hidden;
    }

    /* Hide elements has 'hidden' attribute in IE 10- */
    [hidden] {
      display: none !important;
    }

    ::-webkit-file-upload-button {
      /* stylelint-disable-next-line property-no-vendor-prefix */
      -webkit-appearance: button;
      font: inherit;
    }
  }
`

const typographyMixin = css<StyledProps<IProps>>`
  html {
    font-family: ${({theme}) =>
      theme.typography
        .fontFamily}; /* Prevent scrollbars to overlap content in IE11 and Edge */
    font-size: ${({theme}) =>
      toPercent(
        getValue(theme.typography.htmlFontSize),
        16
      )}; /* Set default tap highlight to be transparent in iOS. */
    font-style: normal;
    color: ${({theme}) => theme.palette.text.dark};
  }

  body {
    ${bodyMixin<IProps>()}
  }

  h1 {
    ${h1Mixin<IProps>()}
  }

  h2 {
    ${h2Mixin<IProps>()}
  }

  h3 {
    ${h3Mixin<IProps>()}
  }

  h4 {
    ${h4Mixin<IProps>()}
  }

  h5 {
    ${h5Mixin<IProps>()}
  }

  h6 {
    ${h6Mixin<IProps>()}
  }

  small {
    ${smallMixin<IProps>()}
  }

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.bold};

    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    caption {
      font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
    }
  }
`

const utilsMixin = css<StyledProps<IProps>>`
  .hidden {
    visibility: hidden;
  }

  .d-none {
    display: none;
  }

  .hidden-sm-down {
    ${({theme}) =>
      mediaBreakpointDown(
        "sm",
        theme.grid.breakpoints,
        `
      display: none;
    `
      )}
  }

  .text-center {
    text-align: center;
  }
`

const CssBaseline: GlobalStyleComponent<
  IProps,
  DefaultTheme
> = createGlobalStyle`
  ${genericResetMixin}

  ${appResetMixin}

  ${typographyMixin}

  ${utilsMixin}
`

export default CssBaseline
