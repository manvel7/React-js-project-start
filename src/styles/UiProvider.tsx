import React, {FunctionComponent} from "react"
import {ThemeProvider} from "styled-components"
import theme from "./theme"
import {CssBaseline} from "./index"

interface IProps {}

const UiProvider: FunctionComponent<IProps> = (props) => {
  const {children} = props

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  )
}

export default UiProvider
