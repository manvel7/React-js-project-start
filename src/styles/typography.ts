import {
  css,
  DefaultTheme,
  FlattenInterpolation,
  StyledProps,
  ThemedStyledProps
} from "styled-components"
import {pxToRem} from "../utils/unit"

const h1Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h1.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h1.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h1.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h1.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h2Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h2.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h2.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h2.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h2.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h3Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h3.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h3.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h3.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h3.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h4Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h4.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h4.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h4.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h4.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h5Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h5.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h5.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h5.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h5.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h6Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h6.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h6.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h6.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h6.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const bodyMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.body.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.body.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.lineHeight,
      theme.typography.htmlFontSize
    )};
`

const smallMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.caption.weight};
  letter-spacing: ${({theme}) =>
    theme.typography.variants.caption.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`
export {
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin,
  bodyMixin,
  smallMixin
}
