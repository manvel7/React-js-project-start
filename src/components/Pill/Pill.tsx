import {ButtonHTMLAttributes} from "react"
import styled, {css, StyledProps} from "styled-components"
import {bodyMixin, smallMixin} from "../../styles/typography"

export type PillSize = "sm" | "md"

export interface IProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  size?: PillSize
  isActive?: boolean
}

const smMixin = css<StyledProps<IProps>>`
  ${smallMixin<IProps>()}
`

const mdMixin = css<StyledProps<IProps>>`
  ${bodyMixin<IProps>()}
`

const activeMixin = css<StyledProps<IProps>>`
  background-color: ${({theme}) => theme.palette.primary};
  color: ${({theme}) => theme.palette.text.white};
`

const Pill = styled.button<IProps>`
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: ${({theme}) => theme.palette.primary};
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;

  ${({isActive}) => (isActive ? activeMixin : "")}

  ${({size}) => {
    switch (size) {
      case "sm":
        return smMixin
      case "md":
      default:
        return mdMixin
    }
  }}
`

Pill.defaultProps = {
  isActive: true
}

export default Pill
