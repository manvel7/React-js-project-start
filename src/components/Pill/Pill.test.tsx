import React from "react"
import {render} from "@testing-library/react"
import {UiProvider} from "../../styles"
import Pill from "./Pill"

describe("Pill", () => {
  describe("Active", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill isActive={true}>Active Pill</Pill>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("Active small 'sm' size", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill isActive={true} size="sm">
            Small Pill
          </Pill>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("Active medium 'md' size", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill isActive={true} size="md">
            Medium Pill
          </Pill>
        </UiProvider>
      )

      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("Not Active", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill isActive={false}>Not Active Pill</Pill>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("Not Active small 'sm' size", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill size="sm" isActive={false}>
            Small Pill
          </Pill>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("Not Active medium 'md' size", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Pill size="md" isActive={false}>
            Medium Pill
          </Pill>
        </UiProvider>
      )

      expect(result.container.firstChild).toMatchSnapshot()
    })
  })
})
