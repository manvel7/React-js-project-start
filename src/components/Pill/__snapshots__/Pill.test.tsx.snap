// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill Active medium 'md' size should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  background-color: #001254;
  color: #FEFEFE;
  font-size: 1.6rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 2.4rem;
}

<button
  class="c0"
>
  Medium Pill
</button>
`;

exports[`Pill Active should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  background-color: #001254;
  color: #FEFEFE;
  font-size: 1.6rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 2.4rem;
}

<button
  class="c0"
>
  Active Pill
</button>
`;

exports[`Pill Active small 'sm' size should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  background-color: #001254;
  color: #FEFEFE;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 1.6rem;
}

.c0 b,
.c0 strong {
  font-weight: 800;
}

<button
  class="c0"
>
  Small Pill
</button>
`;

exports[`Pill Not Active medium 'md' size should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  font-size: 1.6rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 2.4rem;
}

<button
  class="c0"
>
  Medium Pill
</button>
`;

exports[`Pill Not Active should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  font-size: 1.6rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 2.4rem;
}

<button
  class="c0"
>
  Not Active Pill
</button>
`;

exports[`Pill Not Active small 'sm' size should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  border: none;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  display: inline-block;
  padding: 0.4rem 1.6rem;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 1.6rem;
}

.c0 b,
.c0 strong {
  font-weight: 800;
}

<button
  class="c0"
>
  Small Pill
</button>
`;
